Source/WebCore/platform/graphics/haiku/GraphicsContextHaiku.cpp

```diff
@@ -299,29 +356,92 @@ void GraphicsContextHaiku::fillRoundedRectImpl(const FloatRoundedRect& roundRect
     m_view->SetHighColor(oldColor);
 }
 
-void GraphicsContextHaiku::fillPath(const Path& path)
+void GraphicsContextHaiku::fillRectWithRoundedHole(const FloatRect& rect, const FloatRoundedRect& roundedHoleRect, const Color& color)
 {
-    HGTRACE(("fillPath: (--todo print values)\n"));
-    m_view->SetFillRule(fillRule() == WindRule::NonZero ? B_NONZERO : B_EVEN_ODD);
-    m_view->MovePenTo(B_ORIGIN);
+    HGTRACEX(("fillRectWithRoundedHole: [%f:%f] [%f:%f] hole: [%f:%f] [%f:%f]\n",
+        rect.x(), rect.y(), rect.width(), rect.height(),
+        roundedHoleRect.rect().x(), roundedHoleRect.rect().y(), roundedHoleRect.rect().width(), roundedHoleRect.rect().height()));
+    Path path;
+    path.addRect(rect);
 
-    // TODO: Render the shadow (cf shadowAndFillCurrentCairoPath)
-    drawing_mode mode = m_view->DrawingMode();
+    if (!roundedHoleRect.radii().isZero())
+        path.addRoundedRect(roundedHoleRect);
+    else
+        path.addRect(roundedHoleRect.rect());
+
+    WindRule oldFillRule = fillRule();
+    Color oldFillColor = fillColor();
+
+    setFillRule(WindRule::EvenOdd);
+    setFillColor(color);
+
+    if (hasDropShadow()) {
+        HGTRACE(("hasDropShadow begin\n"));
+        const auto shadow = dropShadow();
+        ASSERT(shadow);
+        
+        ShadowBlur contextShadow(*shadow, shadowsIgnoreTransforms());
+        contextShadow.drawInsetShadow(*this, rect, roundedHoleRect);
+        HGTRACE(("hasDropShadow end\n"));
+    }
 
-    if (m_state.fillBrush().pattern())
+    fillPath(path);
+
+    setFillRule(oldFillRule);
+    setFillColor(oldFillColor);
+}
+
+void GraphicsContextHaiku::fillPath(const Path& path)
+{
+    FloatRect rect = path.fastBoundingRect();
+    FloatSize layerSize = getCTM().mapSize(rect.size());
+    FloatRect mappedRect = getCTM().mapRect(rect);
+    HGTRACE(("fillPath: fastBoundingRect = [%f:%f] [%f:%f], mappedRect = [%f %f]\n",
+        rect.x(), rect.y(), rect.width(), rect.height(), mappedRect.x(), mappedRect.y()));
+    BRect pathBounds = BRect(0, 0, rect.width() + 1, rect.height() + 1);
+    BBitmap *pathBitmap = new BBitmap(pathBounds, B_RGBA32, true, false);
+    if(m_painter == nullptr) {
+        m_painter = new BView(pathBounds, "painter", B_FOLLOW_ALL, B_WILL_DRAW);
```

Comment:
creating a bitmap and a view is a costly operation (it spawns a thread in app_server at least). Maybe we want to put some caching in place to recycle the bitmap for multiple draing operations if that is needed.